orca-clouddriver/src/test/java/com/netflix/spinnaker/orca/clouddriver/tasks/manifest/WaitForManifestStableTaskTest.java

```diff
@@ -0,0 +1,383 @@
+/*
+ * Copyright 2020 Google, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.orca.clouddriver.tasks.manifest;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.*;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.netflix.spinnaker.orca.ExecutionStatus;
+import com.netflix.spinnaker.orca.TaskResult;
+import com.netflix.spinnaker.orca.clouddriver.OortService;
+import com.netflix.spinnaker.orca.clouddriver.model.Manifest;
+import com.netflix.spinnaker.orca.pipeline.model.Execution;
+import com.netflix.spinnaker.orca.pipeline.model.Stage;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import org.assertj.core.api.AssertionsForClassTypes;
+import org.junit.jupiter.api.Test;
+import org.junit.platform.runner.JUnitPlatform;
+import org.junit.runner.RunWith;
+
+@RunWith(JUnitPlatform.class)
+final class WaitForManifestStableTaskTest {
+  private static final String UNSTABLE_MESSAGE = "manifest is unstable";
+  private static final String FAILED_MESSAGE = "manifest failed";
+
+  private static final String ACCOUNT = "my-account";
+  private static final String NAMESPACE = "my-namespace";
+  private static final String MANIFEST_1 = "my-manifest-1";
+  private static final String MANIFEST_2 = "my-manifest-2";
+
+  @Test
+  void terminalWhenFailedStable() {
+    OortService oortService = mock(OortService.class);
+    WaitForManifestStableTask task = new WaitForManifestStableTask(oortService, new ObjectMapper());
+
+    Stage myStage =
+        createStageWithManifests(ImmutableMap.of(NAMESPACE, ImmutableList.of(MANIFEST_1)));
+
+    when(oortService.getManifest(ACCOUNT, NAMESPACE, MANIFEST_1, false))
+        .thenReturn(manifestBuilder().stable(true).failed(true).build());
+
+    TaskResult result = task.execute(myStage);
+    AssertionsForClassTypes.assertThat(result.getStatus()).isEqualTo(ExecutionStatus.TERMINAL);
```

Comment:
I see we're using an `AssertionsForClassTypes.assertThat` in one other test to create an assertion for an Optional...when should we use this vs. vanilla `assertThat`?

Reply:
Great question!  I meant to add a comment on the PR for this, thanks for reminding me!  This is effectively a case study of Effective Java item 52 "Use Overloading Judiciously".

The reason is that `ExecutionStatus` implements `CharSequence` and there are two methods in  `Assertions.java` (where I've replaced the return with `U` in both cases as it's not important):
```
static <T extends Comparable<? super T>> U assertThat(T actual)
```
and a:
```
static U assertThat(CharSequence actual)
```

The compiler is tasked with finding the most specific invocation (using the compile-time types) when passed an `ExecutionStatus` but determines that neither of these methods is "most specific".  Per the JLS 15.12.2.5 "The informal intuition is that one method is more specific than another if any invocation handled by the first method could be passed on to the other one without a compile-time error" (which then goes on to get into a bunch of details I don't fully understand).  But I think here it's that:
* We can't replace an arbitrary invocation of the `CharSequence` one with the generic one because a `CharSequence` isn't guaranteed to implement `Comparable`.
* We can't replace invocations of the generic one with the `CharSequence` one because we might have called it with something that's not a `CharSequence`.
Since neither is "more specific" than the other, the compiler can't figure out which method to call and fails.  I think that if the generic one didn't have the bound `extends Comparable<? super T>` this *would* as it would view the `CharSequence` one as more specific and would call that one.

To get around this, AssertJ [suggests using AssertionsForClass](http://joel-costigliola.github.io/assertj/assertj-core.html#ambiguous-compilation-error) which supports the same generic assertions but does not include overloads accepting interfaces and sidesteps this bug.

Reply:
Wow, that's so confusing, thank you for explaining that. My next question was "why does `ExecutionStatus` implement `CharSequence`?" but the answer is, perhaps unsurprisingly, SpEL. :cry:

Reply:
Ha, I hadn't actually dug that far as to ask why it implemented `CharSequence` :cry: :sob: